core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java

```diff
@@ -0,0 +1,94 @@
+ */
+package org.sonarsource.sonarlint.core.container.connected.update;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.annotation.CheckForNull;
+
+import org.sonarqube.ws.WsComponents;
+import org.sonarqube.ws.WsComponents.Component;
+import org.sonarqube.ws.WsComponents.ShowWsResponse;
+import org.sonarqube.ws.WsComponents.TreeWsResponse;
+import org.sonarqube.ws.client.WsResponse;
+import org.sonarsource.sonarlint.core.container.connected.SonarLintWsClient;
+import org.sonarsource.sonarlint.core.util.StringUtils;
+
+public class ModuleHierarchyDownloader {
+  private final SonarLintWsClient wsClient;
+
+  public ModuleHierarchyDownloader(SonarLintWsClient wsClient) {
+    this.wsClient = wsClient;
+  }
+
+  public Map<String, String> fetchModuleHierarchy(String rootProjectKey) {
+    WsResponse response = wsClient.get("api/components/tree.protobuf?qualifiers=TRK,BRC&baseComponentsKey=" + StringUtils.urlEncode(rootProjectKey));
+    try (InputStream stream = response.contentStream()) {
+      TreeWsResponse treeResponse = WsComponents.TreeWsResponse.parseFrom(stream);
+
+      // doesn't include root
+      Map<String, Component> components = treeResponse.getComponentsList().stream().collect(Collectors.toMap(Component::getId, Function.identity()));
+
+      // component -> ancestorComponent. Doesn't include root
+      Map<Component, Component> ancestors = new HashMap<>();
+      for (Component c : treeResponse.getComponentsList()) {
+        ancestors.put(c, components.get(fetchAncestorId(c.getId())));
+      }
+
+      // module key -> path from root project base directory
+      Map<String, String> modulesWithPath = new HashMap<>();
+      modulesWithPath.put(rootProjectKey, "");
+      treeResponse.getComponentsList().forEach(c -> modulesWithPath.put(c.getKey(), findPathFromRoot(c, ancestors)));
+
+      return modulesWithPath;
+
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to load module hierarchy", e);
+    }
+  }
+
+  private String findPathFromRoot(Component component, Map<Component, Component> ancestors) {
+    Component c = component;
+    Path path = Paths.get("");
+
+    do {
+      path = Paths.get(c.getPath()).resolve(path);
+      c = ancestors.get(c);
+    } while (c != null);
+
+    return path.toString();
+  }
+
+  @CheckForNull
+  private String fetchAncestorId(String moduleId) throws IOException {
+    WsResponse response = wsClient.get("api/components/show.protobuf?id=" + StringUtils.urlEncode(moduleId));
+    try (InputStream stream = response.contentStream()) {
+      ShowWsResponse showResponse = WsComponents.ShowWsResponse.parseFrom(stream);
+      return showResponse.getAncestorsList().stream().map(c -> c.getId()).findFirst().orElse(null);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Replace this lambda with a method reference. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)